stage3/buf.C

```diff
@@ -68,7 +68,37 @@ const Status BufMgr::unPinPage(File *file, const int PageNo, const bool dirty) {
 
 }
 
-const Status BufMgr::allocPage(File *file, int &pageNo, Page *&page) {}
+const Status BufMgr::allocPage(File *file, int &pageNo, Page *&page) {
+  if (file == nullptr)
+    return UNIXERR;
+  auto status = file->allocatePage(pageNo);
+  if (status != OK) 
+    return status;
+  
+  int frameNo = -1;
+  //obtain a buffer pool frame
+  status = allocBuf(frameNo);
+  if (status != OK) 
+    return status;
+  
+  //an entry is inserted into the hash table
+  if (hashTable == nullptr)
+    return HASHTBLERROR;
+  status = hashTable->insert(file, pageNo, frameNo);
+  if (status != OK)
+    return status;
+  
+  //The BufDesc class is used to keep track of the state of each frame in the buffer pool.
+  //thus we have to update the correct frame
+  //in the buff table
+  if (bufTable[frameNo] == nullptr) {
```

Comment:
Do you mean `bufTable == nullptr`? 

`BufMgr::bufTable` is an array of `BufDesc` created on heap in `BufMgr` constructor. It makes no sense to test if a `BufDesc` object is equal to `nullptr`.